# Importing Libraries

In [1]:
import pandas as pd

In [2]:
import numpy as np
import os

# Setting Path

In [3]:
path = r'C:\Users\Chaot\02-2024 Instacart Basket Analysis'

In [4]:
path

'C:\\Users\\Chaot\\02-2024 Instacart Basket Analysis'

# Importing Dataframes

In [5]:
df_orders = pd.read_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
df_prod = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'products.csv'), index_col = False)

# Data Consistency Checks

In [7]:
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [8]:
# Creating dataframe
df_test = pd.DataFrame()

In [9]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [10]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [11]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len(df_test[weird]) > 0:
        print(col)

mix


In [12]:
df_test['mix'] = df_test['mix'].astype('str')

# Finding Missing Values

In [13]:
df_prod.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [14]:
#Viewing results for product name
df_nan = df_prod[df_prod['product_name'].isnull() == True]

In [15]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [16]:
df_prod.shape

(49693, 5)

In [17]:
df_prod_clean = df_prod[df_prod['product_name'].isnull() == False]

In [18]:
df_prod_clean.shape

(49677, 5)

# Finding Duplicates

In [19]:
df_dups = df_prod_clean[df_prod_clean.duplicated()]

In [20]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [21]:
df_prod_clean.shape

(49677, 5)

In [22]:
df_prod_clean_no_dupe = df_prod_clean.drop_duplicates()

In [23]:
df_prod_clean_no_dupe.shape

(49672, 5)

# Exercise 4.5 Questions and Answers

## 4.2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [24]:
df_orders.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

The minimum days_since_prior_order is showing 0, which could mean that there might be something wrong here. Either missing the value completely or it's a string instead of an integer.

## 4.3 Check for mixed-type data in your df_ords dataframe.

In [25]:
for col in df_orders.columns.tolist():
  weird = (df_orders[[col]].map(type) != df_orders[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_orders[weird]) > 0:
    print (col)
else :
    print ("No Mix Data")

No Mix Data


## 4.4 If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There are no mixed data types in the df_orders dataframe.

## 4.5 Run a check for missing values in your df_ords dataframe.

In [26]:
df_orders.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

As mentioned after using describe(), the column days_since_prior_order has a total of 206,209 missing values. It is possible that this could be the first time a customer has placed an order and if so, they wouldn't have any order beforehand to count the days from the newest order.

## 4.6 Address the missing values using an appropriate method.

Because the number of missing values is so high, we should consider that replacing the value should not be considered. Instead, we should look and see if these orders are first time orders. Once we've identified the order is new, we can create a new column with it as a boolean data type to mark orders with True or False for them.

## 4.7 Run a check for duplicate values in your df_ords data.

In [27]:
df_orders_dupes = df_orders[df_orders.duplicated()]

In [28]:
df_orders_dupes

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [29]:
df_orders_dupes.shape

(0, 7)

## 4.8 Address the duplicates using an appropriate method.

There are no duplicates within the df_orders dataframe.

## 4.9 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [30]:
#Exporting Products
df_prod_clean_no_dupe.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'products_checked.csv'))

In [31]:
#Exporting Orders
df_orders.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', 'orders_checked.csv'))